File: 04-labs/2026-09-07-lab2-dirty-data/solution/03_figure.R
# ---------------------------------------------------------------
# 03_figure.R -- SOLUTION, lab 2 step 7
#
# Input  : data/processed/trade_clean.csv   (produced by 02_cleaning.R)
# Output : output/figures/oak_destinations.png
#
# This is the script you were given, unchanged. What follows is why
# each choice was made, which is the part you were asked to work out.
# ---------------------------------------------------------------

# --- Dependencies ---
library(here)
library(ggplot2)

# --- Parameters ---
# Named constants rather than numbers buried in the code: when the HS code
# changes, it changes in one place, and the change is visible in a diff.
OAK   <- 440391    # HS code: oak logs
N_TOP <- 5         # destinations shown by name; the rest become "Other"

# --- Load ---
# The output of the previous script, not the raw file. That is what makes
# this a separate stage: re-running it costs nothing, and it can be handed
# to someone who never reads 02_cleaning.
clean <- read.csv(here("data", "processed", "trade_clean.csv"))

# --- 1. Filter ---
oak <- subset(clean, cmdCode == OAK &                  # oak logs only
                     reporterDesc == "France" &        # declared by France
                     flowDesc == "Export")             # outgoing flows

# --- 2. Aggregate ---
# One value per (year, destination). Read the formula as: sum primaryValue,
# for each combination of period and partnerDesc.
by_country <- aggregate(primaryValue ~ period + partnerDesc, data = oak, FUN = sum)

# Rank destinations over the whole period, not year by year: a country that
# is fourth every year should not disappear because it is never first.
totals <- aggregate(primaryValue ~ partnerDesc, data = by_country, FUN = sum)
totals <- totals[order(-totals$primaryValue), ]
top    <- head(totals$partnerDesc, N_TOP)

# Everything outside the top five becomes "Other". Without this the legend
# has sixty entries, the palette recycles colours, and the figure says
# nothing. Re-aggregating afterwards is what makes the Others add up.
by_country$destination <- ifelse(by_country$partnerDesc %in% top,
                                 by_country$partnerDesc, "Other")
graph <- aggregate(primaryValue ~ period + destination, data = by_country, FUN = sum)

# A factor fixes the legend order. Left as text, ggplot2 sorts alphabetically
# and "Other" lands in the middle of the ranking, which reads as a country.
graph$destination <- factor(graph$destination, levels = c(top, "Other"))

# --- 3. Plot ---
# Stacked bars rather than lines: three years make three points, and six
# series on three points is a worse figure than three bars. A stacked bar
# carries the total and its composition in the same object.
ggplot(graph, aes(x = factor(period),              # one bar per year
                  y = primaryValue / 1e6,          # height, in millions
                  fill = destination)) +           # the split inside the bar
  geom_col() +
  labs(title = "French oak log exports (HS 440391)",
       x = NULL, y = "million USD", fill = "Destination")

# --- Output ---
# Into output/figures/, which is regenerable and deletable. Nothing of value
# lives there: if this file is lost, this script rebuilds it.
ggsave(here("output", "figures", "oak_destinations.png"),
       width = 9, height = 5, dpi = 150)

cat("Figure written. Top destination:", as.character(totals$partnerDesc[1]),
    paste0("(", round(totals$primaryValue[1] / 1e6, 1), " M$ over three years)\n"))

File: 04-labs/2026-09-07-lab2-dirty-data/solution/03_figure.py
# ---------------------------------------------------------------
# 03_figure.py -- SOLUTION, lab 2 step 7
#
# Input  : data/processed/trade_clean.csv   (produced by 02_cleaning.py)
# Output : output/figures/oak_destinations.png
#
# This is the script you were given, unchanged. What follows is why
# each choice was made, which is the part you were asked to work out.
# ---------------------------------------------------------------

# --- Dependencies ---
# pandas draws the chart, matplotlib does the drawing. They are two
# packages: python -m pip install matplotlib
from pathlib import Path
import pandas as pd
import matplotlib.pyplot as plt

# --- Parameters ---
ROOT  = Path(__file__).resolve().parents[1]
OAK   = 440391    # HS code: oak logs
N_TOP = 5         # destinations shown by name; the rest become "Other"

# --- Load ---
# The output of the previous script, not the raw file.
clean = pd.read_csv(ROOT / "data" / "processed" / "trade_clean.csv")

# --- 1. Filter ---
oak = clean[(clean.cmdCode == OAK)                     # oak logs only
            & (clean.reporterDesc == "France")         # declared by France
            & (clean.flowDesc == "Export")]            # outgoing flows

# --- 2. Aggregate ---
# pivot_table gives one row per year and one column per destination, which
# is exactly the shape a stacked bar chart wants. fill_value=0 turns "this
# country bought nothing that year" into a zero rather than a hole.
by_country = oak.pivot_table(index="period", columns="partnerDesc",
                             values="primaryValue", aggfunc="sum", fill_value=0)

# Rank over the whole period, not year by year.
totals = by_country.sum().sort_values(ascending=False)
top    = list(totals.index[:N_TOP])

# Everything outside the top five is summed into one column.
graph = by_country[top].copy()
graph["Other"] = by_country.drop(columns=top).sum(axis=1)

# --- 3. Plot ---
# Stacked bars rather than lines: three years make three points.
ax = (graph / 1e6).plot(kind="bar", stacked=True, figsize=(9, 5), width=0.6)
ax.set_title("French oak log exports (HS 440391)")
ax.set_xlabel("")
ax.set_ylabel("million USD")
# The legend outside the axes: inside, it covers the tallest bar.
ax.legend(title="Destination", bbox_to_anchor=(1.02, 1), loc="upper left")
plt.tight_layout()

# --- Output ---
plt.savefig(ROOT / "output" / "figures" / "oak_destinations.png", dpi=150)

print("Figure written. Top destination:", totals.index[0],
      f"({totals.iloc[0]/1e6:.1f} M$ over three years)")
